control-plane/pkg/reconciler/sink/kafka_sink.go

```diff
@@ -161,12 +185,62 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, ks *eventing.KafkaSink) r
 
 func (r *Reconciler) finalizeKind(ctx context.Context, ks *eventing.KafkaSink) error {
 
-	// logger := log.Logger(ctx, "finalize", ks)
+	logger := log.Logger(ctx, "finalize", ks)
+
+	// Get sinks config map.
+	sinksConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get sinks config map %s: %w", r.Configs.DataPlaneConfigMapAsString(), err)
+	}
+
+	logger.Debug("Got sinks config map")
+
+	// Get sinks data.
+	sinks, err := r.GetDataPlaneConfigMapData(logger, sinksConfigMap)
+	if err != nil {
+		return fmt.Errorf("failed to get sinks: %w", err)
+	}
+
+	logger.Debug(
+		"Got sinks data from config map",
+		zap.Any("sinks", (*log.ContractMarshaller)(sinks)),
+	)
+
+	sinkIndex := coreconfig.FindResource(sinks, ks.UID)
+	if sinkIndex != coreconfig.NoResource {
+		coreconfig.DeleteResource(sinks, sinkIndex)
+
+		logger.Debug("Sink deleted", zap.Int("index", sinkIndex))
+
+		// Update the configuration map with the new sinks data.
+		if err := r.UpdateDataPlaneConfigMap(ctx, sinks, sinksConfigMap); err != nil {
+			return err
+		}
+
+		logger.Debug("Sinks config map updated")
+	}
+
+	if ks.GetStatus().Annotations[TopicOwnerAnnotation] == ControllerTopicOwner {
+		topic, err := r.ClusterAdmin.DeleteTopic(ks.Spec.Topic, ks.Spec.BootstrapServers)
+		if err != nil {
+			return err
+		}
+		logger.Debug("Topic deleted", zap.String("topic", topic))
+	}
 
-	// TODO implement finalizer
 	return nil
 }
 
 func incrementGeneration(generation uint64) uint64 {
```

Comment:
This function is cloned in several places, can you move it to core? (in another pr is fine too)

Reply:
Sure in a follow-up.